File: src/tusk/lexer.py
from tusk.token import Token
"""
keywords = [
      
      "then", "elseif", "else",
      "that",
      "times","do","as",
      "what","type",
      "characters","items","all",
      "from","of","by","till",
      "capture",
      "get","post","headers","tson",
      "character","item","number",
      "file", "variable", 
      "with","between",
      "named",
      "for", "can",
      "a", "an", "the","so","to",
      "toall",
      "can","cannot",
      "because",
]

types = ["NUMBER","STRING","BOOL","BOOLEAN","LIST","NOTHING","TSON"]

EFFECTS = [
    "set",
    "print",
    "wait",
    "add","remove","split","replace", 
    "length",
    "input","convert",
    "shell", "python",
    "request",
    "index",
    "read","write","rename",
    "delete",
    "random",
    "import",
]

STRUCTURES = [
    "if","while","function","loop"
]
"""

# May not be the best lexer, but hey- if it works it works and don't touch it, you'll prob mess it up

class Lexer:
    def __init__(self,text, interpreter):
        self.text = text

        self.pos = 0
        self.current_token = None

        self.tokens = []

        self.interpreter = interpreter
      
    def reg(self, name, value):
        self.tokens.append(Token(name.upper(), value, self.interpreter))
        self.ctoken = ""
    
    def classify_tokens(self):
        stuff = self.text.split()

        text = self.text
        reader_pos = 0
        self.ctoken = ""

        in_string = False
        in_comment = False
        hex_count = 0
        in_number = False
        start_quote_type = None     

        for i in "(){}[],;:,":
            text = text.replace(i,f" {i} ")
        text = text.replace("'s "," 's ")
        
        text+="\n"
        while reader_pos < len(text): 
            j = text[reader_pos]
            if in_string:
                if j == start_quote_type:
                    in_string = False
                    self.ctoken = self.ctoken.replace("\\n","\n")
                    self.reg("STRING", self.ctoken)
                else:
                    if start_quote_type=="'" and self.ctoken=="s ":
                        in_string = False
                        self.tokens.append(Token("PROPERTY","'s ",self.interpreter))
                        self.ctoken = ""
                    self.ctoken += j
            elif in_comment:
                if j == "\n":
                    in_comment = False
                    self.ctoken = ""
                else: pass
            elif in_number:
                if j in "0123456789.":
                    self.ctoken += j
                elif j in ["+", "-", "*", "/","*", "%"]:
                    in_number = False
                    self.tokens.append(Token("NUMBER", self.ctoken, self.interpreter))
                    reader_pos -= 1
                elif j in " \t\n":
                    in_number = False
                    self.tokens.append(Token("NUMBER", self.ctoken, self.interpreter))
                    self.ctoken = ""
            elif hex_count > 0:
                if j in "0123456789abcdef":
                    hex_count += 1
                    if hex_count == 7:
                        self.tokens.append(Token("HEX", int("0x" + self.ctoken[1:] + j, 16), self.interpreter))
                        self.ctoken = ""
                        hex_count = 0
                    else:
                        self.ctoken += j
                else:
                    hex_count = 0
            else:
                if j in "(){}[],;:":
                    token_type = {
                        "(": "LEFT_PAR",
                        ")": "RIGHT_PAR",
                        "{": "LEFT_CURLY",
                        "}": "RIGHT_CURLY",
                        "[": "LEFT_SQUARE",
                        "]": "RIGHT_SQUARE",
                        ",": "COMMA",
                        ";": "SEMICOLON",
                        ":": "COLON",
                        "{": "LEFT_CURLY",
                        "}": "RIGHT_CURLY",
                    }[j]
                    self.tokens.append(Token(token_type, j, self.interpreter))
                    self.ctoken = ""
                elif j in ["+", "-", "*", "/","**", "%"]:
                    if text[reader_pos+1].isdigit() and (not text[reader_pos-1].isdigit()): # to allow negative numbers
                        self.ctoken = j
                        in_number = True
                    else:
                        self.tokens.append(Token("OPERATOR", j, self.interpreter))
                elif j == "#":
                    if not (len(text) > reader_pos+6 and all(c.lower() in "0123456789abcdef" for c in text[reader_pos+1:reader_pos+7])):
                        in_comment = True
                        self.ctoken = ""
                    else:
                        hex_count = 1
                        self.ctoken = j
                        print(self.ctoken)
                elif j in ["'", '"']:
                    in_string = True
                    start_quote_type = j
                    self.ctoken = ""
                elif j.isdigit():
                    in_number = True
                    self.ctoken=j
                elif j in " \t\n" or reader_pos == len(text)-1:
                    if reader_pos == len(text)-1 and j not in " \t\n": 
                        self.ctoken += j
                    
                    self.ctoken = self.ctoken.replace(" ","").replace("\n","").replace("\t","")
                    if self.ctoken != "":
                        if self.is_number(self.ctoken):
                            self.tokens.append(Token("NUMBER", self.ctoken, self.interpreter))
                            self.ctoken = ""
                        elif self.ctoken in ["true","false"]:
                            self.tokens.append(Token("BOOL", self.ctoken, self.interpreter))
                            self.ctoken = ""
                        elif self.ctoken == "nothing":
                            self.tokens.append(Token("NOTHING", self.ctoken, self.interpreter))
                            self.ctoken = ""
                        elif self.ctoken in ["and","or","not","contains","in","|","&"]:
                            self.tokens.append(Token("LOGIC", self.ctoken, self.interpreter))
                            self.ctoken = ""
                        elif self.ctoken in ["<", ">", "<=", ">=", "==", "!=","is"]:
                            self.tokens.append(Token("COMPARISION", self.ctoken, self.interpreter))
                            self.ctoken = ""
                        elif self.ctoken in keywords:
                            self.tokens.append(Token("KEYWORD", self.ctoken, self.interpreter))
                            self.ctoken = ""
                        elif self.ctoken in EFFECTS:
                            self.tokens.append(Token("EFFECT",self.ctoken,self.interpreter))
                            self.ctoken=""
                        elif self.ctoken in types:
                            self.tokens.append(Token("TYPE",self.ctoken, self.interpreter))
                            self.ctoken=""
                        elif self.ctoken in STRUCTURES:
                            self.tokens.append(Token("STRUCTURE", self.ctoken, self.interpreter))
                            self.ctoken = ""
                        elif self.ctoken.startswith("#") or self.ctoken.startswith("0x"):
                            self.reg("hex", self.ctoken)
                        elif self.ctoken in ["miliseconds","seconds","minutes","hours","days","weeks","months","years","milisecond","second","minute","hour","day","week","month","year"]:
                            if self.ctoken.endswith("s"):
                                self.ctoken = self.ctoken[:-1]
                            self.tokens.append(Token("TIME", self.ctoken, self.interpreter))
                            self.ctoken = ""
                        elif self.ctoken == "end":
                            self.tokens.append(Token("ENDSTRUCTURE",self.ctoken,self.interpreter))
                            self.ctoken = ""
                        elif self.ctoken in ["return","break"]:
                            self.tokens.append(Token("BREAKSTRUCTURE",self.ctoken,self.interpreter))
                            self.ctoken=""
                        else:
                            if not self.ctoken in " \t\n": 
                                self.tokens.append(Token("IDENTIFIER", self.ctoken, self.interpreter))
                                self.ctoken = ""
                else:
                    self.ctoken += j

            reader_pos += 1

        self.tokens.append(Token("ENDSCRIPT", "", self.interpreter))
        return self.tokens
    
    def is_number(self,string):
        try:
            float(string)
            return True
        except:
            return False

File: src/java/tusan/lexer/Lexer.java
package net.tutla.tums.tusan.lexer;

import net.tutla.tums.tusan.Types;
import net.tutla.tums.tusan.Utils;
import net.tutla.tums.tusan.interpreter.Interpreter;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Lexer {

    public String text;
    private final Interpreter interpreter;

    private int pos = 0;
    private final StringBuilder currentToken = new StringBuilder();
    public List<Token> tokens = new ArrayList<>();

    // Keyword Definitions

    public final List<String> structures = Arrays.asList("if","on","loop","while","on", "function");
    public final List<String> effects = Arrays.asList("print", "set", "wait");
    public final List<String> keywords = Arrays.asList("of", "else", "elseif", "then", "to", "do", "as", "times", "items", "characters", "all", "that");
    public final List<String> timeReprs = Arrays.asList("milliseconds","seconds","minutes","hours","days","weeks","months","years","millisecond","second","minute","hour","day","week","month","year");
    public final List<String> types = Utils.getTypeNames();


    // event mappings are in tusan.Utils

    public Lexer(String text, Interpreter interpreter) {
        this.text = text;
        this.interpreter = interpreter;
    }

    private void register(TokenType name, String value) { // adds token
        tokens.add(new Token(name, value, interpreter));
        currentToken.setLength(0);
    }

    public List<Token> classify() {
        String[] symbols = {"(", ")", "{", "}", "[", "]", ",", ";", ":","="};
        for (String sym : symbols){
            text = text.replace(sym, " " + sym + " ");
        }
        text = text.replace("'s ", " 's ") + "\n";
        text = text.replace("'s ", " 's ") + "\n";

        boolean inString = false;
        boolean inComment = false;
        boolean inNumber = false;
        String startQuoteType = "";


        while (pos < text.length()) {
            char j = text.charAt(pos);

            if (inString) { // string checking thing, i yoinked the code for tusan so its pretty shit so it's not worth touching
                if (String.valueOf(j).equals(startQuoteType)) {
                    inString = false;
                    register(TokenType.STRING, currentToken.toString().replace("\\n", "\n"));
                } else {
                    if (startQuoteType.equals("'") && currentToken.toString().equals("s ")) {
                        inString = false;
                        register(TokenType.PROPERTY, "'s ");
                    }
                    currentToken.append(j);
                }

            } else if (inComment) {
                if (j == '\n') {
                    inComment = false;
                    currentToken.setLength(0);
                }

            } else if (inNumber) {
                if (Character.isDigit(j) || j == '.') {
                    currentToken.append(j);
                } else if ("+-*/%".indexOf(j) != -1) {
                    inNumber = false;
                    register(TokenType.NUMBER, currentToken.toString());
                    pos--;
                } else if (Character.isWhitespace(j)) {
                    inNumber = false;
                    register(TokenType.NUMBER, currentToken.toString());
                }

            } else {
                if ("(){}[],;:=".indexOf(j) != -1) {
                    if (j == '('){
                        register(TokenType.LEFT_PAR, String.valueOf(j));
                    } else  if (j == ')'){
                        register(TokenType.RIGHT_PAR, String.valueOf(j));
                    } else if (j == '{'){
                        register(TokenType.LEFT_CURLY, String.valueOf(j));
                    } else if (j == '}'){
                        register(TokenType.RIGHT_CURLY, String.valueOf(j));
                    } else if (j == '['){
                        register(TokenType.LEFT_SQUARE, String.valueOf(j));
                    } else if (j == ']'){
                        register(TokenType.RIGHT_SQUARE, String.valueOf(j));
                    } else if (j == ','){
                        register(TokenType.COMMA, String.valueOf(j));
                    } else if (j == ';'){
                        register(TokenType.SEMICOLON, String.valueOf(j));
                    } else if (j == ':'){
                        register(TokenType.COLON, String.valueOf(j));
                    } else if (j == '='){
                        register(TokenType.EQUAL, String.valueOf(j));
                    }

                } else if ("+-*/%".indexOf(j) != -1) {
                    if (pos + 1 < text.length() && Character.isDigit(text.charAt(pos + 1)) && (pos == 0 || !Character.isDigit(text.charAt(pos - 1)))) {
                        currentToken.setLength(0);
                        currentToken.append(j);
                        inNumber = true;
                    } else {
                        register(TokenType.OPERATOR, String.valueOf(j));
                    }

                } else if (j == '#') {
                    inComment = true;
                    currentToken.setLength(0);
                } else if (j == '\'' || j == '"') {
                    inString = true;
                    startQuoteType = String.valueOf(j);
                    currentToken.setLength(0);
                } else if (Character.isDigit(j)) {
                    inNumber = true;
                    currentToken.setLength(0);
                    currentToken.append(j);

                } else if (Character.isWhitespace(j) || pos == text.length() - 1) {
                    if (pos == text.length() - 1 && !Character.isWhitespace(j)) {
                        currentToken.append(j);
                    }

                    String tok = currentToken.toString().trim();
                    if (!tok.isEmpty()) {
                        if (tok.matches("\\d+(\\.\\d+)?")) {
                            register(TokenType.NUMBER, tok);
                        } else if (tok.equals("true") || tok.equals("false")) {
                            register(TokenType.BOOL, tok);
                        } else if (tok.equals("=")){
                            register(TokenType.EQUAL, tok);
                        } else if (Arrays.asList("and","or","not","contains","in","||","&&").contains(tok)){
                            register(TokenType.LOGIC, tok);
                        } else if (Arrays.asList(">","<","<=",">=","==","!=","is").contains(tok)){
                            register(TokenType.COMPARISION, tok);
                        } else if (tok.equals("nothing")) {
                            register(TokenType.NOTHING, tok);
                        } else if (types.contains(tok)) {
                            register(TokenType.TYPE, tok);
                        } else if (Arrays.asList("return", "break").contains(tok)) {
                            register(TokenType.BREAKSTRUCTURE, tok);
                        } else if (keywords.contains(tok)) {
                            register(TokenType.KEYWORD, tok);
                        } else if (effects.contains(tok)) {
                            register(TokenType.EFFECT, tok);
                        } else if (structures.contains(tok)) {
                            register(TokenType.STRUCTURE, tok);
                        } else if (tok.equals("end")) {
                            register(TokenType.ENDSTRUCTURE, tok);
                        } else if (Utils.isEventType(tok)) {
                            register(TokenType.EVENT, tok);
                        } else if (timeReprs.contains(tok)) {
                            if (tok.endsWith("s")){
                                tok = tok.substring(0, tok.length() - 1);
                            }
                            register(TokenType.TIME, tok);
                        } else {
                            register(TokenType.IDENTIFIER, tok);
                        }
                    }
                    currentToken.setLength(0);

                } else {
                    currentToken.append(j);
                }
            }
            pos++;
        }

        register(TokenType.ENDSCRIPT, "");
        return tokens;
    }
}
